src/util/expr.cpp

```diff
@@ -539,3 +539,134 @@ const_unique_depth_iteratort exprt::unique_depth_cbegin() const
 { return const_unique_depth_iteratort(*this); }
 const_unique_depth_iteratort exprt::unique_depth_cend() const
 { return const_unique_depth_iteratort(); }
+
+static bool precedence(const exprt &a, const exprt &b)
+{
+  // we use the precendences that most readers expect
+  // (i.e., the ones you learn in primary school),
+  // and stay clear of the surprising ones that C has
+  if((a.id()==ID_mult || a.id()==ID_div) &&
+     (b.id()==ID_plus || b.id()==ID_minus))
+    return true;
+
+  if((a.id()==ID_equal || a.id()==ID_notequal ||
+      a.id()==ID_lt || a.id()==ID_gt ||
+      a.id()==ID_le || a.id()==ID_ge) &&
+     (b.id()==ID_and || b.id()==ID_or))
+    return true;
+
+  return false;
+}
+
+// This turns a multi-ary expression into a string,
+// adding parentheses where indicated by the precedence.
+static std::string as_expr_multi_ary(const exprt &src)
+{
+  bool first=true;
+
+  std::string result;
+
+  for(const auto &op : src.operands())
+  {
+    if(first)
+      first=false;
+    else
+    {
+      result+=' ';
+      result+=id2string(src.id());
+      result+=' ';
+    }
+
+    bool need_parentheses=
+      !op.operands().empty() &&
+      !precedence(op, src);
+
+    if(need_parentheses)
+      result+='(';
+
+    result+=op.as_string();
+
+    if(need_parentheses)
+      result+=')';
+  }
+
+  return result;
+}
+
+static std::string as_expr(const constant_exprt &src)
+{
+  auto type=src.type().id();
+
+  if(type==ID_bool)
+  {
+    if(src.is_true())
+      return "true";
+    else if(src.is_false())
+      return "false";
+    else
+      return src.pretty();
+  }
+  else if(type==ID_unsignedbv || type==ID_signedbv)
+  {
+    return integer2string(*numeric_cast<mp_integer>(src));
+  }
+  else if(type==ID_integer)
+  {
+    return id2string(src.get_value());
+  }
+  else if(type==ID_floatbv)
+  {
+    auto f=ieee_floatt(src);
+    return f.to_ansi_c_string();
```

Comment:
`const auto f` or better yet, `return ieee_floatt(src).to_ansi_c_string();`

That would also allow you to get rid of the curly braces and make code terser.